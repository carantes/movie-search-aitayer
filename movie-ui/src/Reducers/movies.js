import {
  REQUEST_MOVIES,
  RECEIVE_MOVIES,
  RECEIVE_ERROR,
} from "../Actions/types";

export const initialState = {
  loading: false,
  movies: [],
  error: null,
};

function movies(state = initialState, action) {
  const { type, payload, loading, movies, error } = action;
  switch (type) {
    case REQUEST_MOVIES:
      return {
        payload,
        loading,
      };
    case RECEIVE_MOVIES:
      return {
        payload,
        movies,
        loading,
      };
    case RECEIVE_ERROR:
      return {
        error,
        loading,
      };
    default:
      break;
  }
  return state;
}

export default movies;
