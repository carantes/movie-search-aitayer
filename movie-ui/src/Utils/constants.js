export const {
  API_URL = "",
  WAIT_INTERVAL = 300,
  TYPE_COUNT = 3,
} = process.env;

export const headers = {
  Accept: "application/json",
};
