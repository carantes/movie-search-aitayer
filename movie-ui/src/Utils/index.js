export * from "./constants";
export { fetchHelper } from "./fetchHelper";
export { default as mocks } from "./mocks";
