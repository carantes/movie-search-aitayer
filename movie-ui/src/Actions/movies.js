import { REQUEST_MOVIES, RECEIVE_MOVIES, RECEIVE_ERROR } from "./types";
import { fetchHelper } from "Utils";

function requestMovies(keyword) {
  return {
    type: REQUEST_MOVIES,
    payload: keyword,
    loading: true,
  };
}

function receiveMovies(movies, keyword) {
  return {
    type: RECEIVE_MOVIES,
    movies,
    keyword: keyword,
    loading: false,
  };
}

function receiveError(err) {
  return {
    type: RECEIVE_ERROR,
    error: err,
    loading: false,
  };
}

function parseData(movies) {
  return movies.filter(result => result !== null);
}

export const fetchGetMovies = keyword => {
  const apiURL = keyword ? `api/search?keyword=${keyword}` : `api/search`;

  return async dispatch => {
    dispatch(requestMovies(keyword));
    return await fetchHelper(apiURL)
      .then(parseData)
      .then(movies => dispatch(receiveMovies(movies, keyword)))
      .catch(err => dispatch(receiveError(err)));
  };
};
