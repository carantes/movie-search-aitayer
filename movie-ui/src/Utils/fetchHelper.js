import { API_URL, headers } from "./constants";

export const fetchHelper = (url, options = {}) =>
  fetch(`${API_URL}/${url}`, { headers, ...options }).then(res => res.json());
