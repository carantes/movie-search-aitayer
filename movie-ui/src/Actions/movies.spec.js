import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock from "nock";
import { fetchGetMovies } from "./movies";
import { mocks } from "Utils";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("Movies Actions", () => {
  let store;

  beforeEach(() => {
    store = mockStore({});
  });

  afterEach(() => {
    nock.cleanAll();
  });

  it("fetch get movies", async () => {
    nock("http://localhost")
      .get("/api/search")
      .query({ keyword: "avengers" })
      .reply(200, mocks.movies); // Mock reponse code and data

    await store.dispatch(fetchGetMovies("avengers"));
    expect(store.getActions()).toMatchSnapshot();
  });

  it("fetch with error", async () => {
    nock("http://localhost")
      .get("/api/search")
      .reply(400); // Mock reponse code and data

    await store.dispatch(fetchGetMovies());
    expect(store.getActions()).toMatchSnapshot();
  });
});
