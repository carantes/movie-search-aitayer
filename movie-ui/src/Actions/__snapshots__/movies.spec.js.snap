// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies Actions fetch get movies 1`] = `
Array [
  Object {
    "loading": true,
    "payload": "avengers",
    "type": "REQUEST_MOVIES",
  },
  Object {
    "keyword": "avengers",
    "loading": false,
    "movies": Array [
      Object {
        "Poster": "https://m.media-amazon.com/images/M/MV5BNDYxNjQyMjAtNTdiOS00NGYwLWFmNTAtNThmYjU5ZGI2YTI1XkEyXkFqcGdeQXVyMTMxODk2OTU@._V1_SX300.jpg",
        "Title": "The Avengers",
        "Type": "movie",
        "Year": "2012",
        "imdbID": "tt0848228",
      },
      Object {
        "Poster": "https://m.media-amazon.com/images/M/MV5BMTM4OGJmNWMtOTM4Ni00NTE3LTg3MDItZmQxYjc4N2JhNmUxXkEyXkFqcGdeQXVyNTgzMDMzMTg@._V1_SX300.jpg",
        "Title": "Avengers: Age of Ultron",
        "Type": "movie",
        "Year": "2015",
        "imdbID": "tt2395427",
      },
      Object {
        "Poster": "https://m.media-amazon.com/images/M/MV5BMjMxNjY2MDU1OV5BMl5BanBnXkFtZTgwNzY1MTUwNTM@._V1_SX300.jpg",
        "Title": "Avengers: Infinity War",
        "Type": "movie",
        "Year": "2018",
        "imdbID": "tt4154756",
      },
      Object {
        "Poster": "https://m.media-amazon.com/images/M/MV5BMTc5MDE2ODcwNV5BMl5BanBnXkFtZTgwMzI2NzQ2NzM@._V1_SX300.jpg",
        "Title": "Avengers: Endgame",
        "Type": "movie",
        "Year": "2019",
        "imdbID": "tt4154796",
      },
    ],
    "type": "RECEIVE_MOVIES",
  },
]
`;

exports[`Movies Actions fetch with error 1`] = `
Array [
  Object {
    "loading": true,
    "payload": undefined,
    "type": "REQUEST_MOVIES",
  },
  Object {
    "error": [Error: Bad Request],
    "loading": false,
    "type": "RECEIVE_ERROR",
  },
]
`;
